spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/reactive/ReactorLoadBalancerExchangeFilterFunction.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2012-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.client.loadbalancer.reactive;
+
+import java.net.URI;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import reactor.core.publisher.Mono;
+
+import org.springframework.cloud.client.ServiceInstance;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.reactive.function.client.ClientRequest;
+import org.springframework.web.reactive.function.client.ClientResponse;
+import org.springframework.web.reactive.function.client.ExchangeFilterFunction;
+import org.springframework.web.reactive.function.client.ExchangeFunction;
+
+/**
+ * An {@link ExchangeFilterFunction} that uses {@link ReactiveLoadBalancer} to execute
+ * requests against a correct {@link ServiceInstance}.
+ *
+ * @author Olga Maciaszek-Sharma
+ * @since 2.2.0
+ */
+public class ReactorLoadBalancerExchangeFilterFunction implements ExchangeFilterFunction {
+
+	private static Log LOG = LogFactory.getLog(LoadBalancerExchangeFilterFunction.class);
+
+	private final ReactiveLoadBalancer.Factory<ServiceInstance> loadBalancerFactory;
+
+	public ReactorLoadBalancerExchangeFilterFunction(
+			ReactiveLoadBalancer.Factory<ServiceInstance> loadBalancerFactory) {
+		this.loadBalancerFactory = loadBalancerFactory;
+	}
+
+	@Override
+	public Mono<ClientResponse> filter(ClientRequest request, ExchangeFunction next) {
+		URI originalUrl = request.url();
+		String serviceId = originalUrl.getHost();
+		if (serviceId == null) {
+			String msg = String.format(
+					"Request URI does not contain a valid hostname: %s",
+					originalUrl.toString());
+			LOG.warn(msg);
+			return Mono.just(
+					ClientResponse.create(HttpStatus.BAD_REQUEST).body(msg).build());
+		}
+		return choose(serviceId).handle((response, sink) -> {
```

Comment:
Curious why `handle()` here rather than just `map()`?

Reply:
Wanted to be able to handle null `serviceId` and add more logs if it occurred informing the user why the operation failed, but I'm quite new to Reactor, so if you can think of a better way, I will be happy to change.

Reply:
yes, so actually at first I wanted to return a different `onNext` already with the error response there, but then thought this would be easier handled by an exception. In this case, I can switch to `map`. Changing.